Crud update

murraycollingwood opened the pull request.

murraycollingwood committed:
Dumbing down the basic CRUD

Corrected instances of 'Propel' to 'Perpl' throughout the document for consistency.
Removed advanced / complicated elements that were not necessary on this page.
Corrected findPK to findPk

murraycollingwood committed:
Correct the use of !TIP

Removed **Tip**

murraycollingwood committed:
Update documentation for column naming conventions

murraycollingwood committed:
Improve SQL query explanation in documentation

Clarified the explanation of the SQL query and its result.

murraycollingwood committed:
Hide the Custom SQL usage section for now

murraycollingwood committed:
Update comment syntax in documentation

murraycollingwood committed:
Hide Query Termination Methods section

Comment out section on Query Termination Methods.

mringler commented on documentation/03-basic-crud.markdown:
````diff
@@ -5,25 +5,28 @@ title: Basic C.R.U.D. Operations
 ```
 
-The column names used in the `setXXX()` methods correspond to the `phpName` attribute of the `<column>` tag in your schema, or to a CamelCase version of the column name if the `phpName` is not set.
+The column names used in the `setXXX()` methods correspond to the column name in the table.  For example: column `id` will generate `setId()`.  If your column name includes underscores these are removed and the following letter is capitalised, for example column `first_name` will generate `setFirstName()`.  Lastly, you can define the `phpName` attribute of the `<column>` tag in your schema, howev ...
+
+> [!TIP]
+> Be sure that your column names avoid any reserved words for your database.
````
Very nice, ordering this correctly, from general case (uses table name) to special case (change with `phpName`).

Since it applies to all the CRUD methods, I wonder if this should go into its own section at the top, giving an overview of what kind of methods are created and how they are named. It's important to know, yet it doesn't really fit into the "Creating Rows" section.

I'd also like it if the sentence about `phpName` sounded more like a feature, something like "You can decouple column name in database from column identifier used in ORM code by setting a `phpName` attribute in schema.xml"?

murraycollingwood replied:
When your application has 300 tables including 5000 fields, keeping the naming convention simple is always best.  I always avoid 'phpName'.

kasparsatke replied:
Yes, great to provide explanation and examples! 

Regarding the `phpName` attribute - I would like to have it crystal clear that it directly determines the generated php method names. And that the standard is the Upper CamelCase Version of your column name. 

Like e.g.:
`<column name="first_name"/>` will generate `->getFirstName()` and `->setFirstName()` and the corresponding filter methods _normally_. Because of the Upper CamelCase default rule.

_BTW: is there a place to override the default behavior in general? I have only found adding prefixes as documented on the configuration page of he documentation._

But for example:

`<column name="first_name" phpName="GIVENNAME"/>` will generate `->getGIVENNAME()` and `->setGIVENNAME()` and the corresponding filter methods. 

And one can make it as silly as one wants.

I believe that whole generated method naming shebang is something is not clear at all when starting with Propel/Perpl. And this page should be for beginners - right?

So @murraycollingwood could you come up with a concise way of clarifying this in the page? Mine is obviously too long ...

murraycollingwood replied:
Yes - I hear you.  In 18 years of Propel/Perpl I've only used a phpName override once.   It was fine, but to keep it simple I focused on choosing the correct name for the table column which really helped.  I even avoided underscores so that I didn't make mistakes with the camel case rules.

For basic CRUD let's keep it to the camel case auto generated setters, getters and filters etc.  
We will definitely introduce and explain "phpName" when we look at all of the additional schema options.

mringler commented on documentation/03-basic-crud.markdown:
```diff
@@ -33,79 +36,40 @@ INSERT INTO author (first_name, last_name) VALUES ('Jane', 'Austen');
 ### Retrieving by Primary Key ###
 
-The simplest way to retrieve a row from the database, is to use the generated `findPK()` method. It simply expects the value of the primary key of the row to be retrieved.
+The simplest way to retrieve a row from the database, is to use the generated `findPk()` method. This assumes you know the primary key of the record you want to retrieve and also that a primary key identified a unique record.  In the following example we are going to presume that our Author table has three fields, the first being the ID field that contains a unique numeric key.
```
I think "value of the primary key" was correct here. To me, knowing the primary key means to know what column is set as PK (i.e. `id`), not a value in a row.

Primary key, by definition identifies a unique record, I think that part should be removed.

murraycollingwood replied:
Agreed!

mringler commented on documentation/03-basic-crud.markdown:
```diff
@@ -126,9 +90,8 @@ $authors = AuthorQuery::create()
-
->**Tip**`filterByXXX()` is the preferred method for creating queries. It is very flexible and accepts values with wildcards as well as arrays for more complex use cases. See [Column Filter Methods](/documentation/reference/model-criteria.html#column_filter_methods) for details.
+> [!TIP]
+> `filterByXXX()` is the preferred method for creating queries. It is very flexible and accepts values with wildcards as well as arrays for more complex use cases. See [Column Filter Methods](/documentation/reference/model-criteria.html#column_filter_methods) for details.
```
I know you didn't write that, but what does "is the preferred method for creating queries" mean? That's nonsense, isn't it?
As a user, don't tell me what I prefer, you're not the boss of me. Also, preferred over what? And it doesn't actually create queries but filters.

And what does "it's flexible and accepts wildcards and arrays" mean?

murraycollingwood replied:
Agreed!

kasparsatke replied:
Yes, good to have it softened to "common" and actually specified that it is for creating conditions. 

@mringler I assume the lingo is because of the history of the page. 

I think there has been a time when you did a lot of stuff with `->where()` which is not so flexible. And that is reflected here in that now `->filterByXXX()` is "preferred" and more powerful ("flexible") in most cases. Still bad and misleading language. Cos beginners won't know about the project history.

About the arrays: yes one can do this. Although the API could be more transparent and reliable/predictable in some cases (looking at you, `CRITERIA::`). <del>About the wildcards I am also not sure what is meant by that.</del>

Edit: found it. [According to the 1.x docs](https://propelorm.org/Propel/reference/model-criteria.html#column_filter_methods) you can use something like `->filterByTitle('War%')` which in the 2.x documentation is only documented for `->where()` and `->condition()`

mringler commented on documentation/03-basic-crud.markdown:
````diff
@@ -149,20 +112,15 @@ $author = AuthorQuery::create()
 ```
 
->**Tip**Propel provides magic methods for this simple use case. So you can write the above query as:
+The Perpl Query API is very powerful. The next chapter will teach you to use it to add conditions on related objects. If you can't wait, jump to the [Query API reference](/documentation/reference/model-criteria.html).
````
Again, you didn't write that, but my god what a train wreck. First it sounds like they are gratifying themselves ("very powerful"), then they want to "teach" me things (instead of just showing), and apparently they assume I am so excited about the whole thing, I can't even contain myself - and even if I couldn't, shouldn't they tell me, young Padawan, staying in control of your emotions you must, learn lots of important things in this chapter you still have to?
I'm sure it's not meant like this, but boy does it rub me the wrong way.

murraycollingwood replied:
Agreed!

kasparsatke replied:
Not many a disciples found their way to my little documentation planet in a 1000 years but understood little they have. Eager to skip wisdom they have been and to the dark side they were lost.

mringler commented on documentation/03-basic-crud.markdown:
```diff
@@ -235,14 +195,17 @@ AuthorQuery::create()
 
->**Tip**A deleted object still lives in the PHP code. It is marked as deleted and cannot be saved anymore, but you can still read its properties:
+> [!TIP]
+> A deleted object still lives in the PHP code. It is marked as deleted and cannot be saved anymore, but you can still read its properties:
```
B-but how does it get from memory into code? Do I have to re-install Propel to get rid of them???

murraycollingwood replied:
Agreed!  Not applicable for basic CRUD.

mringler replied:
Oh, I am not sure, the remark itself might be helpful? I use deleted objects to create notifications like "Deleted entry for name foo bar (id 12)". It's just phrased awkwardly.

kasparsatke replied:
I believe the original author meant that the deleted object is still "living" in memory and may further be processed. Would be interesting though to know/document for how long exactly.

murraycollingwood replied:
@mringler @kasparsatke 
I would never use or assume that this might be available.  When you delete a record it is gone from the database, and there is an argument that the object in memory should be deleted also to avoid bad coding practices.  
If I need information from a record before I delete, then I copy that information to another var, delete the object, then use my var as necessary.
Assuming a value is still laying around in memory somewhere is very dangerous, for example, what if some future developer determined that when a record is deleted the cached copy should also be destroyed.  Or a future "Perpl garbage collector" automatically removed objects that were flagged as ->isDeleted().

A question: Is this a documented behaviour in Propel?

kasparsatke replied:
It is documented on several occasions not explicitly, but rather implicitly. 
In https://perplorm.github.io/documentation/reference/model-criteria.html#reusing-a-query it says

> By default, termination methods like findOne(), find(), count(), paginate(), or delete() clone the original query. That means that you can reuse a query after a termination method:

In https://perplorm.github.io/documentation/06-behaviors.html you will find the `postDelete()` and `postDeleteQuery()` methods you can override. Also here: https://perplorm.github.io/documentation/05-transactions.html#denormalization-and-transactions
In https://perplorm.github.io/documentation/reference/schema.html#foreign-key-element you have the `ondelete` attribute documented you can set to determine what happens with related objects on deletion.

https://perplorm.github.io/blog/2010/04/29/getting-to-know-propel-1-5-keeping-an-aggregate-column-up-to-date.html again shows the use of `postDelete()`:

> Notice that both the postSave() and the postDelete() receive the current connection object, and use it. This is because these methods are called during a transaction, and the connection object should be the same throughout the whole transaction to let Propel and the database revert the transaction is something wrong occurs.

So, the deleted object still exists to make it 1) available for further manipulation and 2) also to possibly revert a transaction if needed

I would still love to know for how long ...

mringler reviewed (approved):
Nice!

Really good to kick out those sections at the end.

I realize I started to (ab-)use review comments to chat. Except maybe the part about PKs, no change is required - IMO this is already much better than before.

murraycollingwood committed:
Refine explanations in basic CRUD documentation

Clarify descriptions of methods and improve wording for better understanding.

murraycollingwood committed:
Hide section on deleted objects in PHP

Commented out a section about deleted objects in PHP.

murraycollingwood commented:
Thanks @mringler 
Those are really helpful comments and I have made some changes accordingly.
Can you see the additional changes I've made?

mringler commented:
Hey @murraycollingwood, let me know when this is ready to merge. Not sure if @kasparsatke wants to chime in, but I can merge it too.

kasparsatke commented:
@mringler Thanks for the ping! I haven't been notified so I did not see you guys already working on that.
@murraycollingwood I like that you simplified some things, explained others were not obvious and also softened the tone a bit. 

As for the simplification or "dumbing down", as you put it, I believe ommiting some extra information on the basic crud page is good as you did however this information should not be lost. I remember also, that you said you always visit this one blog entry for some specific methods - which of course should also not be needed:

> This is pushing all my buttons...  speaking about documentation, there is
> one blog post that I actually book marked because I go back to it so often.
> https://propelorm.org/blog/2010/05/17/getting-to-know-propel-1-5-when-you-really-need-arrays.html
> 
> [@kasparsatke](https://github.com/kasparsatke) any chance of including this, probably as an additional page
> in the CRUD section.

So I can think of two ways for now to (re)integrate advanced methods:
1) An additional page as you - let's call it "Advanced CRUD" for example
2) Footnotes or collapsible items to provide "on-demand" information yet still keep the basic stuff less cluttered with advanced stuff

What are your thoughts on this?

mringler commented:
@kasparsatke You need to watch the repository if you want to get notified without being tagged directly.

kasparsatke commented on documentation/03-basic-crud.markdown:
```diff
@@ -33,79 +36,40 @@ INSERT INTO author (first_name, last_name) VALUES ('Jane', 'Austen');
->**Tip**For each export method, Propel also provides an import method counterpart. So you can easily populate an object from an array using `fromArray()`, and from a string using any of `fromXML()`, `fromYAML()`, `fromJSON()`, and `fromCSV()`.
-
-There are a lot more useful methods offered by the generated objects. You can find an extensive  list of these methods in the [Active Record reference](/documentation/reference/active-record.html).
-
```
Where should we put that? Like an "Advanced CRUD" page for example?
I mean specifically the extra methods. Or cut it because it's still on the active record reference?

kasparsatke commented on documentation/03-basic-crud.markdown:
````diff
@@ -117,7 +81,7 @@ foreach($authors as $author) {
 ```
 
-To add a simple condition on a given column, use one of the generated `filterByXXX()` methods of the Query object, where `XXX` is a column phpName. Since `filterByXXX()` methods return the current query object, you can continue to add conditions or end the query with the result of the method call. For instance, to filter by first name:
+To add a simple condition on a given column, use one of the generated `filterByXXX()` methods of the Query object, where `XXX` is a column name similar to those we use in the setters and getters. Since `filterByXXX()` methods return the current query object, you can continue to add conditions or end the query with the result of the method call. For instance, to filter by first name:
````
_It actually is_ the value of the `phpName` - regardless if user defined or using the default naming convention. See also my other comment about the phpName.

kasparsatke commented on documentation/03-basic-crud.markdown:
```diff
@@ -326,7 +293,9 @@ $author1 = AuthorQuery::create()->findPk(1);
-When setting the `bulk-load` attribute to true on a [table](/documentation/reference/schema.html#table-element), `findPK()` not only maintains the requested object in the instance pool, but the whole table.
+When setting the `bulk-load` attribute to true on a [table](/documentation/reference/schema.html#table-element), `findPk()` not only maintains the requested object in the instance pool, but the whole table.
+
+-->
```
Where to put that stuff? Advanced CRUD page?

murraycollingwood replied:
@kasparsatke 
@mringler 
Yes - we do need to include all of these extras / advanced options, and we are agreed that they don't really fit on the "Basic CRUD" page.  That's awesome!
So, how do we structure the rest of the documentation?  
I think we need to start a new discussion on this, it's potentially a bigger topic.
Let me look through the other documents we have and I will come up with a master plan and then we can discuss that.  

Are we all agreed?

kasparsatke replied:
a plan sounds good 👍

murraycollingwood committed:
New documentation map

murraycollingwood committed:
Documentation map for Perpl

Expanded the documentation map to include detailed sections and folder structures for better navigation.

murraycollingwood committed:
Documentation map

Updated the documentation map to include a proposed structure alongside the current structure. Added detailed descriptions for each section and reorganized the content for clarity.

murraycollingwood committed:
Format documentation list items for clarity

murraycollingwood committed:
Revise documentation map with new topics and insights

Updated the documentation map with new sections and comments on various topics including active record, inheritance, migrations, relationships, behaviors, and configuration.

murraycollingwood commented:
Sorry - I wanted to put this map in a different branch.  
I can't find any way to withdraw these commits???

murraycollingwood commented:
So I've added a "proposed" documentation map.  This is a temporary document to discuss the overall plan.
I'm comparing the current documents with a proposed set.
To summarise, and this is important for us to agree on: 

I want one set of documents at the top level, covering:
- installation 
- configuration
- basic crud
- query
- active record
- transactions
- logging
- inheritance
- migrations

I would then like a reference folder that documents all of the options available.  The reference would be split into four parts: 
- configuration
- schema
- query
- active record

We should have one last folder for behaviors where we provide sufficient detail about how they work and how to implement them retrospectively.

I'm thinking the Cookbook section has to go - the information there should be covered by the reference detail.

What are your thoughts?

kasparsatke commented:
@murraycollingwood 
> Sorry - I wanted to put this map in a different branch. I can't find any way to withdraw these commits???

It's possible but maybe not so straightforward. And I don't know if you are using git command line or a desktop client to commit and push your stuff. 

In Github desktop e.g. you could squash your last commits for this documentation map into one commit and then create a new branch from this one commit.

If you want to do it with git CLI, here is a manual to do so: https://stackoverflow.com/questions/1628563/move-the-most-recent-commits-to-a-new-branch-with-git/1628584#1628584

Or alternatively, you could create a new discussion in https://github.com/perplorm/perplorm.github.io/discussions and put your suggestions there. It would fit, as it is markdown and HTML anyways and hence render your proposed structure as well (I have already checked out that in a preview). Also, as your proposal at https://github.com/murraycollingwood/perplorm.github.io/blob/crud-update/documentation/00-documentation-map.markdown seems to aim at the overall structure of the documentation, it might well be worth to dicuss it first anyways. If you would put this in a discussion item, you would of course then need to revert your commits to the https://github.com/murraycollingwood/perplorm.github.io/tree/crud-update branch afterwards. 

What do you think?

murraycollingwood committed:
Remove Cookbook section from documentation map

Removed the Cookbook section from the documentation map.

murraycollingwood committed:
Revert commit 1fc108673ab8a99b49b9ae4cd5c5a5284d1afc0b

murraycollingwood commented:
@kasparsatke 
I'm just using the browser interface.
Copilot helped undo the commit.
I've started a separate discussion.

murraycollingwood committed:
Delete documentation/00-documentation-map.markdown
